contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,377 @@
+// snapshot by the volume provider it is bound to the corresponding VolumeSnapshotContent through
+// the VolumeSnapshotSpec
+type VolumeSnapshot struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec represents the desired state of the snapshot
+	// +optional
+	Spec VolumeSnapshotSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status represents the latest observer state of the snapshot
+	// +optional
+	Status VolumeSnapshotStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+type VolumeSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Items           []VolumeSnapshot `json:"items"`
+}
+
+// VolumeSnapshotSpec is the desired state of the volume snapshot
+type VolumeSnapshotSpec struct {
+        // PersistentVolumeClaimName is the name of the PVC being snapshotted
+        // +optional
+        PersistentVolumeClaimName string `json:"persistentVolumeClaimName" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+        // SnapshotDataName binds the VolumeSnapshot object with the VolumeSnapshotContent
+        // +optional
+        SnapshotDataName string `json:"snapshotDataName" protobuf:"bytes,2,opt,name=snapshotDataName"`
+
+        // Name of the VolumeSnapshotClass required by the volume snapshot.
+        // +optional
+        VolumeSnapshotClassName string `json:"storageClassName" protobuf:"bytes,3,opt,name=storageClassName"`
+}
+
+type VolumeSnapshotStatus struct {
+        // CreatedAt is the time the snapshot was successfully created. If it is set,
+	// it means the snapshot was created; Otherwise the snapshot was not created.
+        // +optional
+        CreatedAt *metav1.Time `json:"createdAt" protobuf:"bytes,1,opt,name=createdAt"`
+
+        // AvailableAt is the time the snapshot was successfully created and available
+	// for use. A snapshot MUST have already been created before it can be available.
+        // If a snapshot was available, it indicates the snapshot was created.
+        // When the snapshot was created but not available yet, the application can be
+        // resumed if it was previously frozen before taking the snapshot. In this case,
+        // it is possible that the snapshot is being uploaded to the cloud. For example,
+        // both GCE and AWS support uploading of the snapshot after it is cut as part of
+        // the Create Snapshot process.
+        // If the timestamp AvailableAt is set, it means the snapshot was available;
+        // Otherwise the snapshot was not available.
+        // +optional
+        AvailableAt *metav1.Time `json:"availableAt" protobuf:"bytes,2,opt,name=availableAt"`
+
+        // The time the error occurred during the snapshot creation (or uploading) process
+        // +optional
+        FailedAt *metav1.Time `json:"failedAt" protobuf:"bytes,3,opt,name=failedAt"`
```

Comment:
Let's resuse `VolumeError struct` from https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/storage/types.go
instead of three fields (FailedAt, Reason, Message)

Reply:
Done.